probe_src/probe_py/analysis.py

```diff
@@ -59,59 +70,44 @@ def last(pid: int, exid: int, tid: int) -> Node:
                 target = (pid, exid, op.child_thread_id)
             else:
-                # New process always links to exec epoch 0 and main thread
-                # THe TID of the main thread is the same as the PID
-                target = (op.child_process_id, 0, op.child_process_id)
+                target = (op.child_process_id, -1, 0, 0)
             exec_edges.append((node, first(*target)))
         elif isinstance(op, WaitOp) and op.ferrno == 0 and op.ret > 0:
             # Always wait for main thread of the last exec epoch
             if op.ferrno == 0:
-                target = (op.ret, last_exec_epoch.get(op.ret, 0), op.ret)
+                target = (op.ret, -1, last_exec_epoch.get(op.ret, 0), op.ret)
                 fork_join_edges.append((last(*target), node))
         elif isinstance(op, ExecOp):
             # Exec brings same pid, incremented exid, and main thread
             target = pid, exid + 1, 0
             fork_join_edges.append((node, first(*target)))
-
-    # # Make the main thread exit at the same time as each thread
-    for pid, process in process_tree_prov_log.processes.items():
+            
+    # Make the main thread exit at the same time as each thread
+    for (pid), process in process_tree_prov_log.processes.items():
         for exec_epoch_no, exec_epoch in process.exec_epochs.items():
             for tid, thread in exec_epoch.threads.items():
                 if tid != 0:
-                    fork_join_edges.append((last(pid, exec_epoch_no, tid), last(pid, exec_epoch_no, pid)))
-
-    def node_to_label(node: Node) -> str:
-        inner_label = node[3] if isinstance(node[3], str) else node[3].__class__.__name__
-        return f"{node[0]} {node[1]} {node[2]} {inner_label}"
-
-    def node_to_id(node: Node) -> str:
-        return f"\"node_{node[0]}_{node[1]}_{node[2]}_{id(node[3])}\""
-
-    return "\n".join([
-        "strict digraph {",
-        *[
-            f"  {node_to_id(node)} [label=\"{node_to_label(node)}\"];"
-            for node in nodes
-        ],
-        *[
-            f"  {node_to_id(node0)} -> {node_to_id(node1)} [color=\"green\"];"
-            for node0, node1 in program_order_edges
-        ],
-        *[
-            f"  {node_to_id(node0)} -> {node_to_id(node1)} [color=\"yellow\"];"
-            for node0, node1 in exec_edges
-        ],
-        *[
-            f"  {node_to_id(node0)} -> {node_to_id(node1)} [color=\"red\"];"
-            for node0, node1 in fork_join_edges
-        ],
-        "}",
-    ])
+                    fork_join_edges.append((last(pid, exec_epoch_no, tid), last(pid, exec_epoch_no, 0)))
+
+    process_graph = nx.DiGraph()
+    
+    for node in nodes:
+        # op = [pid, exec_id, tid, op_index]
+        process_graph.add_node(node[4], op = [node[0], node[1],node[2],node[5]])
+    
+    for node0, node1 in program_order_edges:
+        process_graph.add_edge(node0[4], node1[4], color = 'green')
```

Comment:
```suggestion
        process_graph.add_node(node)
    
    for node0, node1 in program_order_edges:
        process_graph.add_edge(node0, node1, color = 'green')
```

I *believe* this will work where `node` is a tuple of `(pid, exid, tid, op_no)`. Networkx lets you use any hashable object as the node.

The list `[1, 2, 3]` is not hashable, because lists are mutable, but the tuple with round brackets `(1, 2, 3)` is hashable, because the tuple is immutable.

```
>>> hash([1, 2, 3])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'list'
>>> hash((1, 2, 3))
529344067295497451
```

Reply:
Yes, tuple works as the node. Made the changes!